Sources/QuaternionModule/Quaternion.swift

```diff
@@ -0,0 +1,365 @@
+/// there exist values for which the Euclidean norm cannot be represented.
+/// Using the infinity norm avoids this problem entirely without significant
+/// downsides. You can access the Euclidean norm using the `length` property.
+/// See `Complex` type of the swift-numerics package for additional details.
+public struct Quaternion<RealType> where RealType: Real & SIMDScalar {
+
+  /// The components of the 4-dimensional vector space of the quaternion.
+  ///
+  /// Components are stored within a 4-dimensional SIMD vector with the scalar part
+  /// first, i.e. representing the most common mathmatical representation that is:
+  ///
+  ///     a + bi + cj + dk
+  @usableFromInline @inline(__always)
+  internal var components: SIMD4<RealType>
+
+  /// A quaternion constructed from given 4-dimensional SIMD vector.
+  ///
+  /// Creates a new quaternion by reading the values of the SIMD vector
+  /// as components of a quaternion with the scalar part first, i.e. in the form of:
+  ///
+  ///     a + bi + cj + dk
+  @usableFromInline @inline(__always)
+  internal init(from components: SIMD4<RealType>) {
+    self.components = components
+  }
+}
+
+// MARK: - Basic Property
+extension Quaternion {
+  /// The real part of this quaternion.
+  ///
+  /// If `q` is not finite, `q.real` is `.nan`.
+  public var real: RealType {
+    @_transparent
+    get { isFinite ? components[0] : .nan }
+
+    @_transparent
+    set { components[0] = newValue }
+  }
+
+  /// The imaginary part of this quaternion.
+  ///
+  /// If `q` is not finite, `q.imaginary` is `.nan` in all lanes.
+  public var imaginary: SIMD3<RealType> {
+    @_transparent
+    get { isFinite ? components[SIMD3(1,2,3)] : SIMD3(repeating: .nan) }
+
+    @_transparent
+    set {
+      components = SIMD4(components[0], newValue.x, newValue.y, newValue.z)
+    }
+  }
+
+  /// The additive identity, with real and *all* imaginary parts zero.
+  ///
+  /// See also:
+  /// -
+  /// - .one
+  /// - .i
+  /// - .infinity
+  @_transparent
+  public static var zero: Quaternion {
+    Quaternion(from: SIMD4(repeating: 0))
+  }
+
+  /// The multiplicative identity, with real part one and *all* imaginary parts zero.
+  ///
+  /// See also:
+  /// -
+  /// - .zero
+  /// - .i
+  /// - .infinity
+  @_transparent
+  public static var one: Quaternion {
+    Quaternion(from: SIMD4(1,0,0,0))
+  }
+
+  /// The imaginary unit.
+  ///
+  /// See also:
+  /// -
+  /// - .zero
+  /// - .one
+  /// - .infinity
+  @_transparent
+  public static var i: Quaternion {
+    Quaternion(imaginary: SIMD3(repeating: 1))
+  }
+
+  /// The point at infinity.
+  ///
+  /// See also:
+  /// -
+  /// - .zero
+  /// - .one
+  /// - .i
+  @_transparent
+  public static var infinity: Quaternion {
+    Quaternion(.infinity)
+  }
+
+  /// The conjugate of this quaternion.
+  @_transparent
+  public var conjugate: Quaternion {
+    Quaternion(from: components.replacing(with: -components, where: [false, true, true, true]))
```

Comment:
I would probably write this as Quaternion(from: components * [1, -1, -1, -1]) (or [-1, -1, -1, 1] if we flip the storage order as I mentioned before).

Reply:
This is definitely easier to read/understand, thanks!